Allow user to specify user-agent header for k8s requests

lwander opened the pull request:
@jimmidyson for #344 

It's not clear to me how to build this project, so this could be totally broken. Is there a getting started guide for developers?

jimmidyson commented:
We need to add a contributing/building guide but you should just be able to run `mvn clean install` & that should be it.

jimmidyson commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Config.java:
```diff
@@ -68,6 +68,8 @@
   public static final String KUBERNETES_ALL_PROXY = "all.proxy";
   public static final String KUBERNETES_NO_PROXY = "no.proxy";
 
+  public static final String KUBERNETES_USER_AGENT = "fabric8-kubernetes-client";
```
Wonder if we can compile in the version number somehow? That would help with debugging & issue tracking. Something like fabric8-kubernetes-client/version would be awesome. Wonder if @iocanel has any idea how we can do that?

iocanel replied:
@jimmidyson: We could have a filtered resource under src/main/resources that would contains ${project.version} this would be substituted during build and could be loaded by Config in a static block.

T

jimmidyson replied:
That sounds perfect.

jimmidyson replied:
Don't mind adding the version in subsequent PR though.

jimmidyson replied:
Added version available as `Version.clientVersion()` so can we add the default user agent as `"fabric8-kubernetes-client/" + Version.clientVersion()`.

jimmidyson commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/HttpClientUtils.java:
```diff
@@ -139,6 +139,16 @@ public Response intercept(Chain chain) throws IOException {
             }
 
+            if (config.getUserAgent() != null && !config.getUserAgent().isEmpty()) {
+                httpClient.interceptors().add(new Interceptor() {
```
Should this be added as a network interceptor target than a normal interceptor?

lwander replied:
I'm not familiar with the difference, this is the only documentation on okHttp interceptors I can find: http://square.github.io/okhttp/3.x/okhttp/

jimmidyson replied:
See more at https://github.com/square/okhttp/wiki/Interceptors.

jimmidyson replied:
Yeah let's add it as a network interceptor please - seen it done in a few places for user agent & seems to be the accepted way. Shame it's not just a simple case of configuring http client but it works...

jimmidyson commented:
Could you add this new option to the readme as well?

Thank you for the contribution. Looking forward to merging this when the few nits are sorted.

lwander commented:
Do you guys normally squash commits before merging?

jimmidyson commented:
I like to review with separate commits (easier to track changes from feedback) & ask for commits to be squashed where appropriate before merge.

lwander commented:
Hey @jimmidyson, I changed the interceptor to a NetworkInterceptor (the link you sent was very helpful in explaining the difference), and I added the version to the UA.

jimmidyson commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/HttpClientUtils.java:
```diff
@@ -139,6 +139,16 @@ public Response intercept(Chain chain) throws IOException {
+                httpClient.networkInterceptors().add(new Interceptor() {
+                    @Override
+                    public Response intercept(Chain chain) throws IOException {
+                        Request agent = chain.request().newBuilder().addHeader("User-Agent", config.getUserAgent()).build();
```
This should just use `header` target than `addHeader` to ensure only single value for the header. Sorry didn't notice that before.

jimmidyson commented:
OK to test

jimmidyson commented:
Happy to merge when changed to `header` & commits squashed please. Thank you!

lwander commented:
All set!

jimmidyson commented:
LGTM [merge]

Thanks!

lwander commented:
Thanks for the help!
